Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +376,120 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
         }
 
         settings.on_message_complete = { parser in
-            let handler = evacuateHTTPDecoder(parser)
-
+            let ctx = evacuateChannelHandlerContext(parser)
+            let handler = ctx.handler as! AnyHTTPDecoder
             handler.state.complete(state: handler.state.dataAwaitingState)
             handler.state.dataAwaitingState = .messageBegin
 
             let trailers = handler.state.currentHeaders.isEmpty ? nil : HTTPHeaders(buffer: handler.state.cumulationBuffer!, headers: handler.state.currentHeaders)
             handler.state.currentHeaders.removeAll(keepingCapacity: true)
             switch handler {
             case let handler as HTTPRequestDecoder:
-                handler.pendingInOut.append(handler.wrapInboundOut(HTTPServerRequestPart.end(trailers)))
+                ctx.fireChannelRead(handler.wrapInboundOut(HTTPServerRequestPart.end(trailers)))
             case let handler as HTTPResponseDecoder:
-                handler.pendingInOut.append(handler.wrapInboundOut(HTTPClientResponsePart.end(trailers)))
+                ctx.fireChannelRead(handler.wrapInboundOut(HTTPClientResponsePart.end(trailers)))
             default:
                 fatalError("the impossible happened: handler neither a HTTPRequestDecoder nor a HTTPResponseDecoder which should be impossible")
             }
             return 0
         }
     }
 
-    public func decoderRemoved(ctx: ChannelHandlerContext) {
-        // Remove the stored reference to ChannelHandlerContext
-        parser.data = UnsafeMutableRawPointer(bitPattern: 0x0000deadbeef0000)
+    private var decoding: Bool = false
```

Comment:
Can we move this up to the other state declarations rather than hiding it just above `channelRead`?